File: myHMMcont.R
myinitHMMcont <- function (States, startProbs = NULL, transProbs = NULL, 
                           emissionParams = NULL)
{
  nStates <- length(States)
  SV <- rep(1 / nStates, nStates)
  names(SV) <- States
  TM <- 0.5 * diag(nStates) + array(0.5 / nStates, c(nStates, nStates))
  dimnames(TM) <- list(from = States, to = States)
  if (is.null(emissionParams)) {
    emissionParams <- lapply(States, function(state) {
      list(mean = 0, sd = 1)
    })
    names(emissionParams) <- States
  }
  if (!is.null(startProbs)) {
    SV[] <- startProbs[]
  }
  if (!is.null(transProbs)) {
    TM[,] <- transProbs[,]
  }
  return(list(
    States = States,
    startProbs = SV,
    transProbs = TM,
    emissionParams = emissionParams
  ))
}


myforwardcont <- function (hmm, observation)
{
  nObservations <- length(observation)
  nStates <- length(hmm$States)
  hmm$transProbs[is.na(hmm$transProbs)] <- 0
  f <- array(NA, c(nStates, nObservations))
  dimnames(f) <- list(states = hmm$States, index = 1:nObservations)
  for (state in hmm$States) {
    mu <- hmm$emissionParams[[state]]$mean
    sigma <- hmm$emissionParams[[state]]$sd
    f[state, 1] <- log(hmm$startProbs[state]) + 
      dnorm(observation[1], mean = mu, sd = sigma, log = TRUE)
  }
  for (k in 2:nObservations) {
    for (state in hmm$States) {
      logsum <- -Inf
      for (previousState in hmm$States) {
        temp <- f[previousState, k - 1] + 
          log(hmm$transProbs[previousState, state])
        logsum <- max(temp, logsum) + log(1 + exp(-abs(temp - logsum)))
      }
      mu <- hmm$emissionParams[[state]]$mean
      sigma <- hmm$emissionParams[[state]]$sd
      f[state, k] <- dnorm(observation[k], mean = mu, sd = sigma, log = TRUE) + 
        logsum
    }
  }
  return(f)
}


mybackwardcont <- function (hmm, observation)
{
  nObservations <- length(observation)
  nStates <- length(hmm$States)
  hmm$transProbs[is.na(hmm$transProbs)] <- 0
  b <- array(NA, c(nStates, nObservations))
  dimnames(b) <- list(states = hmm$States, index = 1:nObservations)
  for (state in hmm$States) {
    b[state, nObservations] = log(1)
  }
  for (k in (nObservations - 1):1) {
    for (state in hmm$States) {
      logsum <- -Inf
      for (nextState in hmm$States) {
        mu <- hmm$emissionParams[[nextState]]$mean
        sigma <- hmm$emissionParams[[nextState]]$sd
        temp <- b[nextState, k + 1] + log(hmm$transProbs[state, nextState]) + 
          dnorm(observation[k + 1], mean = mu, sd = sigma, log = TRUE)
        logsum <- max(temp, logsum) + log(1 + exp(-abs(temp - logsum)))
      }
      b[state, k] <- logsum
    }
  }
  return(b)
}


log_sum_exp <- function(x)
{
  max_x <- max(x)
  max_x + log(sum(exp(x - max_x)))
}


myBWRcont <- function (hmm, obs_list) {
  nOBSseq <- length(obs_list)
  nStates <- length(hmm$States)
  logstart_Y <- matrix(nrow = nOBSseq, ncol = nStates)
  dimnames(logstart_Y) <- list(index = 1:nOBSseq, states = hmm$States)
  TransitionMatrix <- hmm$transProbs
  TransitionMatrix[,] <- 0
  logxi_list <- setNames(vector("list", nStates), hmm$States)
  logxi_list <- lapply(logxi_list, function(dummy) setNames(vector("list", nStates), hmm$States))
  EmissionList <- hmm$emissionParams
  EmissionList[] <- lapply(EmissionList, function(y) { y[] <- 0; y })
  loggamma_list <- setNames(vector("list", nStates), hmm$States)
  for (idx in 1:nOBSseq) {
    observation <- obs_list[[idx]]
    nObservations <- length(observation)
    f <- myforwardcont(hmm, observation)
    b <- mybackwardcont(hmm, observation)
    likelihood <- f[hmm$States[1], nObservations]
    for (state in hmm$States[-1]) {
      j <- f[state, nObservations]
      likelihood <- max(likelihood, j) + log(1 + exp(-abs(likelihood - j)))
    }
    for (state in hmm$States) {
      for (nextState in hmm$States) {
        for (i in 1:(nObservations - 1)) {
          logxi_list[[state]][[nextState]] <- c(logxi_list[[state]][[nextState]], 
            f[state, i] + log(hmm$transProbs[state, nextState]) + 
                  dnorm(observation[i + 1], 
                        mean = hmm$emissionParams[[nextState]]$mean, 
                        sd = hmm$emissionParams[[nextState]]$sd, log = TRUE) + 
                  b[nextState, i + 1] - likelihood)
        }
      }
    }
    for (state in hmm$States) {
      loggamma_list[[state]] <- c(loggamma_list[[state]], 
                               f[state, ] + b[state, ] - likelihood)
    }
    for (state in hmm$States) {
      logstart_Y[idx, state] <- f[state, 1] + b[state, 1] - likelihood
    }
  }
  for (state in hmm$States) {
    for (nextState in hmm$States) {
      TransitionMatrix[state, nextState] <- log_sum_exp(logxi_list[[state]][[nextState]]) - 
        log_sum_exp(unlist(logxi_list[[state]]))
    }
  }
  TransitionMatrix <- exp(TransitionMatrix)
  for (state in hmm$States) {
    loggamma_shifted <- loggamma_list[[state]] - max(loggamma_list[[state]])
    if (sum(exp(loggamma_shifted)) > 0) {
      EmissionList[[state]]$mean <- sum(exp(loggamma_shifted) * 
                                          unlist(obs_list)) / 
        sum(exp(loggamma_shifted))
      EmissionList[[state]]$sd <- sqrt(sum(exp(loggamma_shifted) * 
                                             unlist(obs_list)^2) / 
                                         sum(exp(loggamma_shifted)) - 
                                         (EmissionList[[state]]$mean)^2)
    } else {
      EmissionList[[state]]$mean <- 0
      EmissionList[[state]]$sd <- 1
    }
    if (EmissionList[[state]]$sd < 1e-4) {
      EmissionList[[state]]$sd <- 1e-4
    }
  }
  return(list(
    TransitionMatrix = TransitionMatrix,
    EmissionList = EmissionList,
    startVec = exp(apply(logstart_Y, 2, log_sum_exp))))
}


myBWcont <- function (hmm, obs_list, maxIterations = 100, delta = 1e-9, 
                      pseudoCount = 0)
{
  tempHmm <- hmm
  tempHmm$transProbs[is.na(hmm$transProbs)] <- 0
  diff <- c()
  for (i in 1:maxIterations) {
    bw <- myBWRcont(tempHmm, obs_list)
    TM <- bw$TransitionMatrix
    EL <- bw$EmissionList
    SV <- bw$startVec
    TM[!is.na(hmm$transProbs)] <- TM[!is.na(hmm$transProbs)] + pseudoCount
    SV[!is.na(hmm$startProbs)] <- SV[!is.na(hmm$startProbs)] + pseudoCount
    TM <- TM / rowSums(TM)
    SV <- SV / sum(SV)
    d <- sqrt(sum((tempHmm$transProbs - TM)^2)) + 
      sqrt(sum((unlist(tempHmm$emissionParams) - unlist(EL))^2)) + 
      sqrt(sum((tempHmm$startProbs - SV)^2))
    diff <- c(diff, d)
    tempHmm$transProbs <- TM
    tempHmm$emissionParams <- EL
    tempHmm$startProbs <- SV
    if (d < delta) {
      break
    }
  }
  tempHmm$transProbs[is.na(hmm$transProbs)] <- NA
  tempHmm$startProbs[is.na(hmm$startProbs)] <- NA
  return(list(hmm = tempHmm, difference = diff))
}

File: myHMMcont.py
import numpy as np
import copy


def myinitHMMcont(states, start_probs=None, trans_probs=None, emission_params=None):
    n_states = len(states)
    sv = np.full(n_states, 1.0 / n_states)
    tm = 0.5 * np.eye(n_states) + (0.5 / n_states) * np.ones((n_states, n_states))
    if emission_params is None:
        emission_params = [{'mean': 0.0, 'sd': 1.0} for _ in states]
    if start_probs is not None:
        sv = np.array(start_probs, dtype=float)
    if trans_probs is not None:
        tm = np.array(trans_probs, dtype=float)
    return {'states': states, 'start_probs': sv, 'trans_probs': tm, 'emission_params': emission_params}


def log_sum_exp(x):
    max_x = np.max(x)
    return max_x + np.log(np.sum(np.exp(x - max_x)))


def log_norm_pdf(x, mu, sd):
    return -0.5 * (np.log(2 * np.pi * sd**2) + ((x - mu)**2) / (sd**2))


def myforwardcont(hmm, observation):
    n_obs = len(observation)
    n_states = len(hmm['states'])
    trans = np.nan_to_num(hmm['trans_probs'], nan=0.0)
    f = np.empty((n_states, n_obs))
    for i in range(n_states):
        mu = hmm['emission_params'][i]['mean']
        sd = hmm['emission_params'][i]['sd']
        f[i, 0] = np.log(hmm['start_probs'][i]) + log_norm_pdf(observation[0], mu, sd)
    for k in range(1, n_obs):
        for i in range(n_states):
            logsum = -np.inf
            for j in range(n_states):
                temp = f[j, k-1] + np.log(trans[j, i])
                logsum = max(temp, logsum) + np.log1p(np.exp(-abs(temp - logsum)))
            mu = hmm['emission_params'][i]['mean']
            sd = hmm['emission_params'][i]['sd']
            f[i, k] = log_norm_pdf(observation[k], mu, sd) + logsum
    return f


def mybackwardcont(hmm, observation):
    n_obs = len(observation)
    n_states = len(hmm['states'])
    trans = np.nan_to_num(hmm['trans_probs'], nan=0.0)
    b = np.empty((n_states, n_obs))
    b[:, n_obs-1] = 0.0
    for k in range(n_obs-2, -1, -1):
        for i in range(n_states):
            logsum = -np.inf
            for j in range(n_states):
                mu = hmm['emission_params'][j]['mean']
                sd = hmm['emission_params'][j]['sd']
                temp = b[j, k+1] + np.log(trans[i, j]) + log_norm_pdf(observation[k+1], mu, sd)
                logsum = max(temp, logsum) + np.log1p(np.exp(-abs(temp - logsum)))
            b[i, k] = logsum
    return b


def myBWRcont(hmm, obs_list):
    n_seq = len(obs_list)
    n_states = len(hmm['states'])
    logstart_Y = np.empty((n_seq, n_states))
    TM = np.zeros((n_states, n_states))
    logxi_list = [[[] for _ in range(n_states)] for _ in range(n_states)]
    loggamma_list = [[] for _ in range(n_states)]
    for idx, obs in enumerate(obs_list):
        n_obs = len(obs)
        f = myforwardcont(hmm, obs)
        b = mybackwardcont(hmm, obs)
        likelihood = log_sum_exp(f[:, n_obs-1])
        for i in range(n_states):
            for j in range(n_states):
                for t in range(n_obs-1):
                    logxi_list[i][j].append(
                        f[i, t] + np.log(hmm['trans_probs'][i, j]) +
                        log_norm_pdf(obs[t+1], hmm['emission_params'][j]['mean'], hmm['emission_params'][j]['sd']) +
                        b[j, t+1] - likelihood
                    )
        for i in range(n_states):
            gamma_vals = f[i, :] + b[i, :] - likelihood
            loggamma_list[i].extend(gamma_vals.tolist())
        for i in range(n_states):
            logstart_Y[idx, i] = f[i, 0] + b[i, 0] - likelihood
    for i in range(n_states):
        denom = log_sum_exp(np.array([xi for row in logxi_list[i] for xi in row]))
        for j in range(n_states):
            num = log_sum_exp(np.array(logxi_list[i][j]))
            TM[i, j] = np.exp(num - denom)
    flat_obs = [x for seq in obs_list for x in seq]
    EM_params = []
    for i in range(n_states):
        vals = loggamma_list[i]
        weights = np.exp(np.array(vals) - max(vals))
        mean = np.dot(weights, flat_obs) / weights.sum()
        var = np.dot(weights, np.square(flat_obs)) / weights.sum() - mean**2
        sd = np.sqrt(var) if var > 1e-8 else 1e-4
        EM_params.append({'mean': mean, 'sd': sd})
    start_vec = np.exp([log_sum_exp(logstart_Y[:, i]) for i in range(n_states)])
    return {'TransitionMatrix': TM, 'EmissionParams': EM_params, 'startVec': start_vec}


def myBWcont(hmm, obs_list, maxIterations=100, delta=1e-9, pseudoCount=0):
    tempHmm = copy.deepcopy(hmm)
    mask_trans = ~np.isnan(hmm['trans_probs'])
    mask_sv = ~np.isnan(hmm['start_probs'])
    diff = []
    for _ in range(maxIterations):
        bw = myBWRcont(tempHmm, obs_list)
        TM = bw['TransitionMatrix']
        EM = bw['EmissionParams']
        SV = bw['startVec']
        TM[mask_trans] += pseudoCount
        SV[mask_sv] += pseudoCount
        TM = TM / TM.sum(axis=1)[:, None]
        SV = SV / SV.sum()
        d = np.sqrt(np.sum((tempHmm['trans_probs'] - TM)**2)) + \
            np.sqrt(np.sum((np.array([p['mean'] for p in tempHmm['emission_params']]) - np.array([p['mean'] for p in EM]))**2 +
                           (np.array([p['sd'] for p in tempHmm['emission_params']]) - np.array([p['sd'] for p in EM]))**2)) + \
            np.sqrt(np.sum((tempHmm['start_probs'] - SV)**2))
        diff.append(d)
        tempHmm['trans_probs'] = TM.copy()
        tempHmm['emission_params'] = EM
        tempHmm['start_probs'] = SV.copy()
        if d < delta:
            break
    tempHmm['trans_probs'][~mask_trans] = np.nan
    tempHmm['start_probs'][~mask_sv] = np.nan
    return {'hmm': tempHmm, 'difference': diff}
